### ISSUE TITLE
Possible issue in wettingVelocityFactor (wetting and drying scheme)

### ISSUE THREAD

**Issue by gcapodag** [E3SM-Project/E3SM#5809]
The damping parameter `wettingVelocityFactor` (created in [this](https://github.com/E3SM-Project/E3SM/blob/110bd91147b6337926b3224ba54a3340df4ab275/components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F#L305) routine) in MPAS-Ocean's wetting and drying scheme is defined as a global field that has a value for each edge in the mesh, although I believe this may cause issues in parallel because a process might overwrite a value that it should not: assume cell1 is in proc 1 and cell0 is in proc 0 and they share edge `iEdge` owned by cell1, so `iEdge` is in the halo edges of proc 0. Assume cell0 is drying up and water is flowing from cell0 to cell1 through `iEdge`. Since cell1 owns `iEdge`, it will initialize the value of `wettingVelocityFactor` to zero. Then a check on the flux is done, but since cell1 is not drying up, `wettingVelocityFactor(iEdge)`  is not modified so it stays at a zero value. cell0 on the other hand, is drying up (the flux is leaving cell0 towards cell1 through `iEdge`), so it will change `wettingVelocityFactor(iEdge)` locally to something non zero. A halo update on  `wettingVelocityFactor` would overwrite this non-zero value for proc 0 and reset it to zero. 

It is possible that this behavior is showing in the compass `drying slope test case`  with 1 km resolution, ramp and single layer, because running for 12 hours, the results are not BFB for the normal velocity comparing a serial run with one using 10 processes. It is interesting though that the results are BFB when using 2,3,...,9 processes. Maybe the process partitioning is not such that the above behavior is triggered in those case.

**Comment by xylar** [comment]
@gcapodag, have you actually seen this be a problem?  To me, it looks like `wettingVelocityFactor` is computed on each processor iterating over `nCellsAll`, meaning over not just the owned cells but also the halo cells:
https://github.com/E3SM-Project/E3SM/blob/110bd91147b6337926b3224ba54a3340df4ab275/components/mpas-ocean/src/shared/mpas_ocn_wetting_drying.F#L370
This should mean that both proc 0 and proc 1 would redundantly calculate the same `wettingVelcoityFactor(iEdge)` as the same value as long as there has been a prior halo update for the various input fields (`normalVelocity`, `layerThickEdgeFlux`, `layerThicknessCur`, `layerThicknessProvis`, etc.).  If there is a problem, I would think it would be with a missing or overly restrictive halo update for one or more of those input quantities, rather than for `wettingVelocityFactor` itself.

A way to see if the problem you are identifying is related to what you think would be to add a halo update for `wettingVelocityFactor` and see if that fixes the issue at 1 km resolution.

**Comment by xylar** [comment]
My reading of that bit of code is that we iterate over all cells in both the owned cells and the halo, then we iterate over all edges on those cells.  If there is an outgoing flux from that cell (`normalVelocity(k, iEdge) * edgeSignOnCell(i, iCell) < 0.0_RKIND`), that flux get added to the outward divergence.  That computation should be the same for an owned cell on proc 0 and a halo cell on proc 1 because the halo includes all the edges on halo cells.  I don't think that's the part where you think there could be issues, but correct me if I'm wrong.

Then, both processors compute `wettingVelocityFactor` on all edges on all cells, owned or halo.  They should agree on `layerThickness` in all cells whether they are owned or halo, and they should also agree on whether the flux is in or out (the sign of `normalVelocity(k, iEdge) * edgeSignOnCell(i, iCell)`) whether the cell and edge are owned or halo.  So I don't see how they would get different values for `wettingVelocityFactor(k, iEdge)` depending on who owns the cell.  It seems to me that there is enough information in the halo for both cell 0 and cell 1 to agree on the `wettingVelocityFactor` for a given edge regardless of who owns it.  I was initially concerned that there might be an issue with edges on the outer boundary of the halo but even there I no longer see why there would be disagreement between the processors.

**Comment by xylar** [comment]
Perhaps part of the confusion is the strategy in the algorithm.  @cbegeman would be able to describe it better than I can.  But here is my best attempt.

In order to prevent iteration, rather than computing the full divergence, the algorithm is conservative and doesn't take into account incoming thickness fluxes at edges.  It assumes a cell will dry up if there is enough outgoing flux to dry it up.  This means that each cell effectively "owns" the edges with outgoing flux and it gets to determine the `wettingVelocityFactor` for those edges.  This "ownership" should not be related to which processor the cell is owned by (or the edge, for that matter), since all processors should agree on whether the flux is outgoing from a given cell.  The `wettingVelocityFactor` can only be set based on properties of the cell that that flux is coming out of, and even if that cell is a halo cell for one processor and an owned cell for another, both processors should agree on the sign of the flux at a given edge and the total outward flux in the given cell so they should, in my view, agree on the `wettingVelocityFactor`.

@gcapodag and @cbegeman, I'm hoping you can pick up the conversation from here as I think you two might be able to work out the details of whether there is an issue better than I can.  But so far I don't see why there would be one.

**Comment by gcapodag** [comment]
Hi @xylar, I think I am on the same page as you now, since each process loops over all cells (owned plus halo) as you mentioned, I now see that the computation is indeed redundant if all the fields involved have been properly updated. Since I am looking into W/D for LTS, I am going to look into this non-BFB situation for the compass test case @cbegeman, so no need for you to spend time on this. I will write here as soon as I find out what is causing it. Thanks for the convo @xylar !

**Comment by gcapodag** [comment]
@xylar do you want me to rename this issue (or create a new one) to mention the non-BFB issue in the title or should we just leave everything as it is?

**Comment by xylar** [comment]
@gcapodag, I think it might be cleanest to open a new issue and close this one, though you can reference the discussion here if you think it's relevant.

**Comment by gcapodag** [comment]
Sounds good! Thanks!

**Comment by cbegeman** [comment]
@gcapodag Thanks for looking into this. As we discussed on slack, I also don't know why the `wettingVelocityFactor` would be the reason for decomp test (1 vs 10 proc) failures. I'll keep an eye out for another issue.
